### DISCUSSION TITLE
New FreeCAD Devs: tell us how you've learned from our documentation how to setup a dev environment to hack on FreeCAD

### DISCUSSION THREAD

**Discussion by luzpaz** [FreeCAD/FreeCAD#17595] [Code contribution discussions]
This discussion is a way to get feedback from interested devs who want to contribute to FreeCAD and want to setup an appropriate dev environment. 
* Are new devs successful with this process? 
* Can they easily find the appropriate documentation for their needs (OS/Platform, 3rd-party dependencies, clear and concise docs)
* What is working?
* What needs to be improved?
* Any useful hacks that helped lessen the learning curve to building FreeCAD and coding in FreeCAD ?
* Any suggestion for the workflow to be modified or augmented ?

Please document your step-by-step process of setting up your dev environment and learning process to compile and test FreeCAD. Along with deibugging (git bisect; valgrind; strace) issues. 

Thanks for your participation!

Edit: 
Opened a ticket in the Developer documentation to track feedback from this discussion:
* https://github.com/FreeCAD/DevelopersHandbook/issues/105

**Comment by JohnOCFII** [comment]
Since @oursland created the well documented Conda build process, I've been regularly building FreeCAD on my Mac.  When asked to test simple code changes, I've been able to hand-edit them into my local files for testing.  Parts of the process where I'm still a bit gun-shy:

- Actually submitting a PR to FreeCAD
- Learning the proper way to test other's code

The https://freecad.github.io/DevelopersHandbook/gettingstarted/ is a good start.  Some areas are pretty thin, so, for example it'd be great to get an expanded "newbie friendly" section for PR submission.  Granted, much of this is really about becoming familiar with `Git` and `Github` operations, but a "See Spot, See Spot Run" level guide would be helpful.

I'd love to see the FreeCAD build instructions better align with (and perhaps prefer) the Conda build instructions.  There are so many ways to set up a build environment, and while that's great -- I think for newbies coming into the project, it might really streamline CMake issues, etc., if folks start with the Conda build processes.

It has been 38 years since I got a degree in "Quantitative methods and Computer Science" and 33 years since I was considered a "programmer," but beyond the current little bit of Mac QA testing that I do, I'd love to take the next step and test and submit simple fixes.

**Reply by bofdahof** [reply]
I absolutely want a unified build for Mac. Chasing dependencies via brew has been a never-ending "there's always one more little problem to solve" exercise. FreeCAD hanging onto QT5 is not helping. Git Modules are a pain as well.

I see weeklies etc are produced by conda stuff, but what about QT6?

I'll keep my fingers crossed and look forward to others' experiences.

**Comment by howie-j** [comment]
My struggles has been with finding and installing the required dependencies. It seems like the required packages are documented a bit here and there, and the wiki ([https://wiki.freecad.org/Compile_on_Linux](https://wiki.freecad.org/Compile_on_Linux)) has mostly been slightly outdated. I tried to keep the Fedora section updated, but that became hard a year ago when Fedora 39 stopped supporting Qt5.

IMO there should be **_one_** source where all dependencies are are listed, preferably somewhere in the official FreeCAD git repo. All wiki and developer documentation should point to this list. All pr's that introduces or removes a dependency should also update this list.

Furthermore, i think **_one_** system-independent build system should be recommended for all devs, preferably conda or whatever the CI/CD uses.

**Comment by longrackslabs** [comment]
I've set up the dev environment in Linux Mint and Mac OS (M4 Arm).

And to be honest, both times it was a complete nightmare. Dependencies, certain versions of libraries, and the main thing was even disabling different features with build flags to try to avoid dependency and include problems. It's like certain libraries and headers are hard coded in.

I think I probably spa a full day eight hours trying to get things to build on my new Mac.  I got there, but it was not fun.

My first thought was to run on docker, which I was able to get a bill to run, but there's x11 display problems so that didn't really work all that great.

So once I got my unknown working, set up and code tasks I save those away and hope I never have to set up a new environment  again.

I get this feeling that I am a rare breed doing development on a ARM Mac.  I'd love to make this better but the build process is so complicated. I'm afraid I'd break something.

Again, this is just my experience as a relatively new FreeCAD contributor

**Reply by JohnOCFII** [reply]
> I get this feeling that I am a rare breed doing development on a ARM Mac. I'd love to make this better but the build process is so complicated. I'm afraid I'd break something.

Did you use the Pixi build instructions in the Developer's Handbook?  I'm not really doing development but building is quite easy on Arm Mac with this process.  I have seen folks getting assistance getting debugging setup via the IDE and the Pixi build process being discussed in the Discord #Development channel too.

https://freecad.github.io/DevelopersHandbook/

**Reply by tarman3** [reply]
[pixi](https://freecad.github.io/DevelopersHandbook/gettingstarted/#pixi) instruction is terrible
A lot of steps missed and some them not obvious

There is no step on how to download sources and what to do with them
Probably the author of the text never tried to repeat what was written

[forum](https://forum.freecad.org/viewtopic.php?p=840388#p840388)

**Reply by JohnOCFII** [reply]
> There is no step on how to download sources and what to do with them

That's in the `Setting up for Development [#]`(https://freecad.github.io/DevelopersHandbook/gettingstarted/#setting-up-for-development) section of the Developer's Handbook.  I'd agree, there are some sections that could be re-ordered, but I found if I took the time to read through the whole thing, I was able to get it to work for me.  I've been doing Pixi builds for quite some time now.  Like everything, there are days or weeks where there are bumps in the road, but for the most part, my daily routine consists of:

```
git pull
cd build
pixi run configure
pixi run build
```

**Reply by oursland** [reply]
> [pixi](https://freecad.github.io/DevelopersHandbook/gettingstarted/#pixi) instruction is terrible A lot of steps missed and some them not obvious

These commands are complete and tested in VMs.  Many newcomers and experienced developers find the guide effective for adopting `pixi` for builds.  Which steps do you believe are missing?

**Comment by longrackslabs** [comment]


**Reply by onekk** [reply]
You should have problems when working for a different Linux distributions, as behind them there are different policies that lead to different source of problems:

- libraries naming, it is very different if you work with a Debian based, RedHat based, Arch based or Gentoo distributions only to name the most used flavours.
- libraries are often split in pieces, as example it is not rare to find a libxxx, libxxx-dev, libxxx-doc in Debian and derivatives and ususlly most distributions separate the documentation from the library, but it is not a rule, for small libraries even on Debian you could find only a package, for big libraries often the library itself could come splitted in pieces and so the dev and the doc thing.
- Some distribution, have some strict ruels about licenses so it is not infrequent to have some libraries that contains different licensed portions be stripped in free and non-free packages (This could happen on Debain derivatives)

It is generally complicated to write a build recipe that will work on several distributions, even Debian and Ubuntu has some quirks that forbid to mix libraries if they are not taylored to be shared between Ubuntu and Debian as usually Ubuntu is based on the "testing branch" of Debian, so the "stable" release of the two similar distro could have different library versions, and a different build recipe.
 
Sadly there is not a reference Linux distribution (and probably it will never be as the Linux Foundation is not caring of this detail).

Kind regards
Carlo D.

**Comment by longrackslabs** [comment]
Thanks for the reply.  I understand that there were diffference in LInux.
But my main challenges were hard coded library includes that were ignored
by -D flags.  That was a huge frustration.  The FEM stuff and others, I
just could not seem to get to turn off.  So this was as much difficutly on
my LInux Mint as it has been on my new MacBookPro M4 ARM.

The cmake / build process seems pretty brittle in geenral.  I got
everything working on both platforms, but it was a major chore.

I appreciate how complex and brittle such an old and complicated code base
must be.

thanks,
G

**Reply by JohnOCFII** [reply]
> The cmake / build process seems pretty brittle in geenral. I got everything working on both platforms, but it was a major chore.

The Pixi builds take care of all that for you -- including regularly updating the underlying CMAKE configs, etc.  It really is the way forward.  If you have issue, pop in to the Discord where the build maintainers hang out, and they'll get you sorted.

https://freecad.github.io/DevelopersHandbook/gettingstarted/

**Comment by depthoffocus** [comment]
I have just started doing builds again on an intel Mac and the Pixi builds are fine. I have been able to fiddle about with the Mac builds to try to diagnose the recent menu item movement weirdness and it has gone OK.

**Comment by depthoffocus** [comment]
By the way, is the documentation still correct that Visual Studio Community 2019 is required? Because it can no longer be downloaded without a Visual Studio/MSDN subscription:

https://visualstudio.microsoft.com/vs/older-downloads/

I gather the requirement comes from conda-forge but that has moved on to the 2022 version:
https://conda-forge.org/news/2025/06/11/moving-to-vs2022/

**Reply by loopless** [reply]
Visual Studio 2022 works fine

**Comment by depthoffocus** [comment]
Additionally, is there anywhere I can read about the parts of the build that e.g. generate the AppImage or (in my case) the macOS `.app` build?

I was interested in understanding how the `.app` was built in particular, and I can see that the Rattler build seems to do it, but it's not clear to me if this is part of the documentation anywhere; I have not been able to find anything.
